File: leetcode/108/BinarySearchTree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {Integer[]} nums
# @return {TreeNode}
def sorted_array_to_bst(nums)
    return nums if nums.empty?
    the_actual_function(nums, 0, nums.length-1)
end

def the_actual_function(nums, s, e)
    if s <= e
        mid = (s+e) / 2
        node = TreeNode.new(nums[mid])
        node.left = the_actual_function(nums, s, mid - 1)
        node.right = the_actual_function(nums, mid+1, e)
        node
    end
end

File: leetcode/108/BinarySearchTree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode SortedArrayToBST(int[] nums) {
        if (nums == null || nums.Length == 0)
            return null;
        
        return BuildBST(nums, 0, nums.Length - 1);
    }
    
    private TreeNode BuildBST(int[] nums, int startIndex, int endIndex)
    {
        TreeNode currentRoot = null;
        int currentRootIndex = endIndex + (startIndex - endIndex) / 2;
        
        if (startIndex <= endIndex)
        {
            currentRoot = new TreeNode(nums[currentRootIndex]);
            currentRoot.left = BuildBST(nums, startIndex, currentRootIndex - 1);
            currentRoot.right = BuildBST(nums, currentRootIndex + 1, endIndex);
        }
        
        return currentRoot;
    }
}